### ISSUE TITLE
Please clarify the sense of `tasks.named { ... }` opposed to `tasks.matching { ... }`, also in release notes, resp. actually make them behave different

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#28347]
### Current Behavior

I know that you usually do not modify published release notes, but the release notes for 8.6 are plainly wrong.
They state that `tasks.matching { it.name.contains("pack") }.configureEach { ... }` does realize all tasks which is not true.
And that due to that `tasks.named { it.contains("pack") }.configureEach { ... }` was introduced in 8.6 to mitigate that problem.
But these statements are plainly wrong.

`tasks.matching { it.name.contains("pack") }.configureEach { ... }` did never and still does not realize any additional tasks.

`tasks.matching { it.name.contains("pack") }.forEach { ... }` causes all tasks to be realized.
But so does `tasks.named { it.contains("pack") }.forEach { ... }`.

`tasks.matching { it.name.contains("pack") }.all { ... }` causes all tasks to be realized.
But so does `tasks.named { it.contains("pack") }.all { ... }`.

`val packs by tasks.registering { dependsOn(tasks.matching { it.name.contains("pack") }) }` causes all tasks to be realized.
But so does `val packs by tasks.registering { dependsOn(tasks.named { it.contains("pack") }) }`.

I would have hoped `tasks.named { ... }` can actually work without realizing tasks in situations where `tasks.matching` cannot, for example when using `forEach` or `all` or `dependsOn`, as there only the name should be necessary and not the whole element like with `matching`.
But as it is currently, as far as I have seen both behave exactly identical and you just have two ways now to do exactly the same with exactly the same sideeffects.

### Expected Behavior

The release notes should not state the false fact that confuses users reading that.
Optimally, the `named { ... }` would be reworked to not force-realize any unmatched tasks even if eager APIs like `all` or `forEach` or `dependsOn` are used, as this should be totally unnecessary and then actually be a significant improvement over `tasks.matching`.

### Steps to Reproduce

n/a

### Gradle version

8.6

**Comment by Vampire** [comment]
Here is another user wondering about the same: https://discuss.gradle.org/t/question-regarding-8-6-release-notes-tasks-named/47865

**Comment by boaznahum** [comment]
@mgroth0 I think any attempt to use TaskContainer iteration "default" method , like size(), iterator() causes realization of tasks.

So in your example, even simple line like(groovy):

```groovy
def filtered = tasks.named { itn ->
        println("$itn:\tNamed referenced in `named`")
        return itn.contains("subPack")
    }

println "filtered=${filtered.size()}"
```

So the only way to correctly iterate this Iterbale is by
```
configureEach
```
But the later can't be used in register context
```groovy
tasks.register("packs") { it->

    println("$it:\tregistering")
    
    filtered.configureEach { f ->
        t.dependsOn f
    }
}
```

throws an exception:
```
A problem occurred configuring root project 'kotlin-registering-gradle'.
> Could not create task ':packs'.
   > DefaultTaskCollection#configureEach(Action) on task set cannot be executed in the current context.

```

Seems like a deadend or desisgn  hole  :worried:

**Comment by Vampire** [comment]
Yes, as I said, it behaves exactly identical to `matching` as far as I have seen.

**Comment by ov7a** [comment]
Thank you for raising this, we are investigating the issue.

**Comment by jbartok** [comment]
@Vampire, you are correct, the release notes are wrong. I've written [a fix](https://github.com/gradle/gradle/tree/jb/clarify-named-collection-method-in-release-notes) and will try to fix the already released documentation.

In a nutshell, yes, when you just call `matching()` or `named()`, they behave the same way, until you iterate the filtered task container they produce. Once you do that `matching()` has unpleasant side-effects, because it triggers the creation of ALL lazy tasks in the original container, regardless if they have been filtered out or not. `named()` doesn't have this flaw.

Please let me know if you still have issues with the fixed version of the release notes.

**Comment by Vampire** [comment]
@jbartok as I said, in 8.6 both behave identical while I hoped for the behavior / difference you described which unfortunately is **not** the case.

For example:
- do `gw init --use-defaults`
- add to the build script
  ```kts
  tasks.configureEach { println("configured $path") }
  tasks.named { it == "help" }.all {}
  ```
- run `gw -m`

Output:
```plain
> Configure project :
configured :help
configured :projects
configured :tasks
configured :properties
configured :dependencyInsight
configured :dependencies
configured :buildEnvironment
configured :components
configured :model
configured :dependentComponents
configured :outgoingVariants
configured :resolvableConfigurations
configured :javaToolchains
configured :prepareKotlinBuildScriptModel
configured :init
configured :wrapper
configured :kotlinDslAccessorsReport
:help SKIPPED
```

Same if you change the build script content to
```kts
tasks.configureEach { println("configured $path") }
val foo by tasks.registering {
    dependsOn(tasks.named { it == "help" })
}
```
and do `gw foo`.
Output:
```plain
12:14:48: Executing 'foo'...

configured :foo
configured :help
configured :projects
configured :tasks
configured :properties
configured :dependencyInsight
configured :dependencies
configured :buildEnvironment
configured :components
configured :model
configured :dependentComponents
configured :outgoingVariants
configured :resolvableConfigurations
configured :javaToolchains
configured :prepareKotlinBuildScriptModel
configured :init
configured :wrapper
configured :kotlinDslAccessorsReport

> Task :help

Welcome to Gradle 8.6.
[...}
> Task :foo
```

**Comment by Vampire** [comment]
Also same with 8.7-rc-2.
I can not see any difference in behavior between using `matching` or `named`.

**Comment by jbartok** [comment]
@Vampire, you are correct in everything you said, I'm sorry that it took me so long to reach the same conclusion. In my defense, we have tests for the behavior we aimed for, but apparently we shot ourselves in the foot with them, because they are broken in subtle ways... That's not really an excuse, just saying. We will strive to fix the situation as soon as we are able.

**Comment by rkrisztian** [comment]
My experience is the following:

```groovy
// EAGER (configures all tasks)
tasks.matching { it.name.contains('myTask') }.forEach {}

// LAZY
tasks.matching { it.name.contains('myTask') }.configureEach {}

// EAGER (configures all tasks)
tasks.named { it.contains('myTask') }.all {}

// LAZY
tasks.named { it.contains('myTask') }.configureEach {}
```

*Edit:* Furthermore, we get do better for `findByPath` than just telling not to use it:

```groovy
// EAGER
tasks.findByPath(':myTask').configure {}

// LAZY
tasks.matching { it.path == ':myTask' }.configureEach {}
```

**Comment by Vampire** [comment]
Yes, we already reached agreement, that `named {}` is behaving exactly like `matching {}` regarding laziness, even though it should not. :-)
